# 4.5 Data Consistency Checks

In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os

In [2]:
# Reletive path
path = '../'

# Loading dataframe orders.csv
df_ords = pd.read_csv(
    os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv')
)

In [3]:
# Import the “products.csv
df_prods = pd.read_csv(
    os.path.join(path, '02 Data', 'Original Data', 'products.csv'), 
    index_col = False
)

In [45]:
# Making sure that the “products.csv” imported.
df_prods.isnull().sum()

product_id       0
product_name     0
aisle_id         0
department_id    0
prices           0
dtype: int64

In [12]:
# Records containing nulls
df_nan = df_prods[df_prods['product_name'].isnull() == True]
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [ ]:
## to see mean
#df.describe()
## to see median
#df_prods.median()

## Replace N/A values
# df['column with missings'].fillna(mean value, inplace=True)

In [13]:
# Cleanin products data frame
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

# df_prods.dropna(inplace = True)

# df_prods.dropna(subset = [‘product_name’], inplace = True)

In [14]:
df_prods_clean.shape

(49677, 5)

In [47]:
# Command will look for full duplicates within the dataframe
df_dups = df_prods_clean[df_prods_clean.duplicated()]
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [48]:
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

df_prods = df_prods_clean_no_dups
df_prods.shape

(49672, 5)

# The task 4.5 Exercises

In [17]:
# 2. Run the df.describe() function on your df_ords dataframe. Using your new knowledge about how to interpret the output of this function, share in a markdown cell whether anything about the data looks off or should be investigated further.
# Tip: Keep an eye on min and max values!  

df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_prior_order  
count        3.421083e+06       3.421083e+06            3.214874e+06  
mean         2.776219e+00       1.345202e+01            1.111484e+01  
std          2.046829e+00       4.226088e+00            9.206737e+00  
min          0.000000e+00       0.000000e+00            0.000000e+00  
25%          1.000000e+00       1.000000e+01            4.000000e+00  
50%          3.000000e+00       1.300000e+01            7.000000e+00  
75%          5.000000e+00       1.600000e+01            1.500000e+01  
max          6.000000e+00       2.300000e+01            3.000000e+01

Observation: Nothing suspicious, it looks like it has already been cleaned. 
days_since_prior_order : 3, order_hour_of_day 23, order_dow: 6

Described data stats:

| Operation | order_id    | user_id     | order_number | order_dow | order_hour_of_day | days_since_prior_order |
|-----------|-------------|-------------|--------------|-----------|--------------------|------------------------|
| max       | 3.421083e+06 | 2.062090e+05 | 1.000000e+02 | 6         | 23                 | 30                     |



In [22]:

df_prods.describe()

product_id      aisle_id  department_id        prices
count  49672.000000  49672.000000   49672.000000  49672.000000
mean   24850.349775     67.762442      11.728942      9.993282
std    14340.705287     38.315784       5.850779    453.615536
min        1.000000      1.000000       1.000000      1.000000
25%    12432.750000     35.000000       7.000000      4.100000
50%    24850.500000     69.000000      13.000000      7.100000
75%    37268.250000    100.000000      17.000000     11.100000
max    49688.000000    134.000000      21.000000  99999.000000

In [27]:
df_prods[df_prods['prices'] == 99999]

product_id           product_name  aisle_id  department_id   prices
33666       33664  2 % Reduced Fat  Milk        84             16  99999.0

Observations: 
2 % Reduced Fat Milk Having suspicious price

| operation | order_id   | user_id   | order_number | order_dow | order_hour_of_day | days_since_prior_order |
|-----------|------------|-----------|--------------|-----------|--------------------|------------------------|
| id        | 33666      | 33664     | 2 % Reduced Fat Milk | 84 | 16 | 99999.0         |


In [28]:
# 3. Check for mixed-type data in your df_prods dataframe.

# Defining function to check: if the data types in all columns are mixed
def mixed_type_check(df):
    for col in df.columns.tolist():
      weird = (df[[col]].applymap(type) != df[[col]].iloc[0].apply(type)).any(axis = 1)
      if len (df[weird]) > 0:
        print (col)

mixed_type_check(df_prods)

In [29]:
# 4. If you find mixed-type data, fix it. The column in question should contain observations of a single data type.
mixed_type_check(df_ords)

In [30]:
## Converting a column’s data type from numeric to string in df_prods
# df_prods['product_name'] = df_prods['product_name'].astype('str')

/var/folders/1t/82zy_nkj13dg65kz31p835s00000gn/T/ipykernel_16741/2178115741.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_prods['product_name'] = df_prods['product_name'].astype('str')


In [44]:
# 5.1 Run a check for missing values in your df_ords dataframe.
df_ords.isnull().sum()

Unnamed: 0                0
order_id                  0
user_id                   0
eval_set                  0
order_number              0
orders_day_of_week        0
order_hour_of_day         0
days_since_prior_order    0
dtype: int64

In [33]:
df_ords.shape

(3421083, 8)

In [43]:
df_ords[df_ords['days_since_prior_order'].isnull()]

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, eval_set, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

5.2 In a markdown cell, report your findings and propose an explanation for any missing values you find.

Observation: Thare is null values in column days_since_prior_order. Looks like this is a first orders of each partiqular user. By the defenition of this column this should be zero values.

In [42]:
# 6.1 Updating missing zeros in 'days_since_prior_order' column.
df_ords['days_since_prior_order'].fillna(0, inplace=True)

6.2 In a markdown cell, explain why you used your method of choice.

This methon do not use addition memory to update N/A values.

In [46]:
# 7.1 Run a check for duplicate values in your df_ords data.

# Command will look for full duplicates within the dataframe
df_ords_dups = df_ords[df_ords.duplicated()]
df_ords_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, eval_set, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

In [49]:
# Command will look for full duplicates within the dataframe
df_prods_dups = df_prods[df_prods.duplicated()]
df_prods_dups

Empty DataFrame
Columns: [product_id, product_name, aisle_id, department_id, prices]
Index: []

7.2 In a markdown cell, report your findings and propose an explanation for any duplicate values you find.

Obcervation: No duplicate vlues found in df_ords.

8.1 Address the duplicates using an appropriate method.
In a markdown cell, explain why you used your method of choice.

Were used command:
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

This command is a simple and not memory consuming method. 

In [50]:
# Export df_prods_clean_no_dups
df_prods.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_cleaned.csv'))

# Step 9: Export final df_ords 
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_cleaned.csv'), index = False)